clang/include/clang/Driver/Options.td

```diff
@@ -2900,7 +2900,7 @@ defm clangir : BoolFOption<"clangir",
   PosFlag<SetTrue, [], [ClangOption, CC1Option], "Use the ClangIR pipeline to compile">,
   NegFlag<SetFalse, [], [ClangOption, CC1Option], "Use the AST -> LLVM pipeline to compile">,
   BothFlags<[], [ClangOption, CC1Option], "">>;
-def emit_cir : Flag<["-"], "emit-cir">, Visibility<[CC1Option]>,
+def emit_cir : Flag<["-"], "emit-cir">, Visibility<[ClangOption, CC1Option]>,
```

Comment:
@erichkeane @AaronBallman @MaskRay 

I had to re-add the ClangOption here. `lib/Driver/Driver.cpp` builds the `Action` framework out based on input to the driver. So to properly orchestrate the outputs you need a driver flag to tell it that, e.g., we are emitting a `.cir` textual output. e.g. `-fsyntax-only` exists to chop off the frontend at the right point and `-emit-llvm` seems intentionally to have been made a driver flag to orchestrate `.ll` or `.bc` output.

Notably, `-emit-llvm-bc` does the wrong thing in all cases. `clang -Xclang -emit-llvm-bc hello.c` does the wrong thing and emits an executable. `clang -Xclang -emit-llvm -S hello.c` also does the wrong thing and emits a file named `hello.s` that is actually llvm bitcode.

Reply:
> Notably, -emit-llvm-bc does the wrong thing in all cases. clang -Xclang -emit-llvm-bc hello.c does the wrong thing and emits an executable. clang -Xclang -emit-llvm -S hello.c also does the wrong thing and emits a file named hello.s that is actually llvm bitcode.

The description of #91140 might be useful.

You need `-S` or `-c` to disable linking. Then `-c -Xclang -emit-cir` can be used to override the assumed action passed by clangDriver to cc1.

Reply:
>  Then `-c -Xclang -emit-cir` can be used to override the assumed action passed by clangDriver to cc1.

It would be incorrect to only use `-Xclang -emit-cir -c` as it would output a `.o` file for a textual `.cir` file. I've plugged in ClangIR to the action building system. `-c` requests actions through the assembler. We're stopping with ClangIR via `-emit-cir` after the `Compile` phase. 

> -flto modifies -c to emit a bitcode file and -S to emit a textual IR file. -fcir can be modeled like -flto.

That's an inaccurate comparison. `-fcir` is supposed to only change from using codegen to cirgen&lowering. The output shouldn't change. `-emit-cir` is supposed to emit a file with the suffix `.cir`. The *driver* has to be told the preferred output type to tell the action system to emit a `.cir` suffixed file.

Reply:
@MaskRay are you okay with this approach? Also, if this is exposed as a driver flag, should we have a test in `clang/test/Driver` that it emits the expected -cc1 option?

Reply:
Fangrui is out for a few weeks: https://github.com/llvm/llvm-project/pull/104899#pullrequestreview-2248433784

Reply:
To restate the argument here in favor of using a `-emit-cir` frontend flag:

* [This line shows](https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/Driver.cpp#L4862) the chunk of driver code responsible for mapping the requested emission to the output file. Without `-emit-cir` as a frontend flag we instead get `.o`, `.s` or some other incorrect output based on the `-c`, `-S` flags you pass to the frontend. 
* @MaskRay  previously argued that we should use `-fclangir` to change the output. But this flag is orthogonal to the output. It merely enabled the ClangIR pipeline and, by design, does not change the output. We need *some* flag to tell the driver what the output kind is supposed to be.